lago/config.py

```diff
@@ -14,88 +14,216 @@
-    'libvirt_password': '',
-}
-
-
-def _get_environ():
-    return os.environ
-
-
-def _get_from_env(key):
-    return _get_environ()['LAGO_%s' % key.upper()]
-
-
-def _get_from_files(paths, key):
-    config = ConfigParser.ConfigParser(defaults=DEFAULTS)
-    config.read(paths)
-    try:
-        return config.get('lago', key)
-    except ConfigParser.Error:
-        raise KeyError(key)
-
-
-def _get_from_dir(path, key):
-    config_files = glob.glob(os.path.join(path, '*.conf'))
-    return _get_from_files(config_files, key)
-
-
-def _get_providers():
-    return [
-        _get_from_env,
-        functools.partial(
-            _get_from_files,
-            [_USER_CONFIG],
-        ),
-        functools.partial(
-            _get_from_dir,
-            _SYSTEM_CONFIG_DIR,
-        ),
+import re
+from collections import defaultdict
+from io import StringIO
+from warnings import warn
+
+import configparser
+from xdg import BaseDirectory as base_dirs
+from .constants import CONFS_PATH
+from .utils import argparse_to_ini
+
+
+def _get_configs_path():
+    paths = []
+    xdg_paths = [
+        path for path in base_dirs.load_config_paths(
+            'lago', 'lago.conf'
+        )
     ]
-
-
-_cache = {}
-_GET_DEFAULT = object()
-
-
-def get(key, default=_GET_DEFAULT):
-    if key in _cache:
-        return _cache[key]
-
-    for provider in _get_providers():
-        try:
-            val = provider(key)
-            _cache[key] = val
-            return val
-        except KeyError:
-            pass
-
-    # Nothing was found
-    if default is _GET_DEFAULT:
-        raise KeyError(key)
-    else:
-        return default
+    paths.extend(CONFS_PATH)
+    paths.extend(reversed(xdg_paths))
+    return paths
+
+
+def get_env_dict(root_section):
+    """Read all Lago variables from the environment.
+    The lookup format is:
+      LAGO_VARNAME - will land into 'lago' section
+      LAGO__SECTION1__VARNAME - will land into 'section1' section, notice
+      the double '__'.
+
+    Returns:
+        dict: dict of section configuration dicts
+
+    Examples:
+        >>> os.environ['LAGO_GLOBAL_VAR'] = 'global'
+        >>> os.environ['LAGO__INIT__REPO_PATH'] = '/tmp/store'
+        >>>
+        >>> config.get_env_dict()
+        {'init': {'repo_path': '/tmp/store'}, 'lago': {'global_var': 'global'}}
+
+    """
+    env_lago = defaultdict(dict)
+    regex = r'((?P<nosect>^{0}_(?!_))|(?P<sect>^{0}__))(?P<name>.*$)'.format(
+        re.escape(root_section.upper())
+    )
+    decider = re.compile(regex)
+    for key, value in os.environ.iteritems():
+        match = decider.match(key)
+        if match:
+            if not match.group('name') or not value:
+                warn(
+                    'empty environment variable definition:'
+                    '{0}, ignoring.'.format(key)
+                )
+            elif match.group('nosect'):
+                env_lago[root_section][match.group('name').lower()] = value
```

Comment:
maybe assign the groups to variables to make more readable this part of the code, something like:

```python
_, name, section, global = [word.lower() for word in match.groups()]
```

It feels a bit strange too that you have the sect group in the regexp but then use split to extract it from the name, maybe you should try to get it already from the regexp (if you haven't tried, if it gets too complicated it's not worth it imo)